File: packages/reflex-sdk/src/compiler.js
/**
 * Reflex Compiled Instinct & Edge Binary Runtime for Node.js, Bun, Cloudflare Workers, and Browser.
 * Parses and executes portable .reflex models (RFX1 format) in <20us.
 * Zero external dependencies.
 */

import { SemanticVectorEncoder } from "./encoder.js";
import { Choice, Noul, Score, DecisionResult } from "./primitives.js";

// Precomputed IEEE 802.3 CRC32 lookup table
const CRC32_TABLE = new Uint32Array(256);
for (let i = 0; i < 256; i++) {
  let c = i;
  for (let j = 0; j < 8; j++) {
    c = (c & 1) ? (0xedb88320 ^ (c >>> 1)) : (c >>> 1);
  }
  CRC32_TABLE[i] = c >>> 0;
}

/**
 * Computes the 32-bit CRC checksum of a byte buffer.
 * @param {Uint8Array} bytes
 * @returns {number} Unsigned 32-bit integer
 */
export function crc32(bytes) {
  let crc = 0xffffffff;
  for (let i = 0; i < bytes.length; i++) {
    crc = CRC32_TABLE[(crc ^ bytes[i]) & 0xff] ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

function sigmoid(z) {
  const clamped = Math.max(-30.0, Math.min(30.0, z));
  return 1.0 / (1.0 + Math.exp(-clamped));
}

function softmax(logits, temperature = 1.0) {
  if (!logits || logits.length === 0) return [];
  const temp = Math.max(0.01, temperature);
  const scaled = logits.map((x) => x / temp);
  const maxL = Math.max(...scaled);
  const exps = scaled.map((x) => Math.exp(Math.max(-30.0, Math.min(30.0, x - maxL))));
  const sumExps = exps.reduce((acc, x) => acc + x, 0);
  if (sumExps <= 0.0) {
    return logits.map(() => 1.0 / logits.length);
  }
  return exps.map((x) => x / sumExps);
}

/**
 * Portable, self-contained compiled decision model evaluating in <20us.
 */
export class CompiledInstinct {
  /**
   * @param {Object} options
   * @param {string} options.name
   * @param {"choice"|"noul"|"score"} options.decisionType
   * @param {string[]} options.options
   * @param {Record<string, number[]>} options.weights
   * @param {Record<string, number>} options.biases
   * @param {number} [options.temperature=1.0]
   * @param {Object} [options.metrics]
   * @param {Object} [options.spec]
   * @param {string} [options.compiledAt]
   * @param {string} [options.version="1.0"]
   */
  constructor({
    name,
    decisionType,
    options = [],
    weights = {},
    biases = {},
    temperature = 1.0,
    metrics = null,
    spec = null,
    compiledAt = null,
    version = "1.0",
  }) {
    this.name = name;
    this.decisionType = decisionType;
    this.options = options;
    this.weights = weights;
    this.biases = biases;
    this.temperature = Math.max(0.01, temperature);
    this.metrics = metrics || { accuracy: 1.0, brier_score: 0.0, ece: 0.0 };
    this.spec = spec;
    this.compiledAt = compiledAt;
    this.version = version;
    this.encoder = new SemanticVectorEncoder();
  }

  /**
   * Loads and verifies a .reflex binary model from a Uint8Array or Buffer.
   * @param {Uint8Array|ArrayBuffer|Buffer} buffer
   * @returns {CompiledInstinct}
   */
  static fromBinary(buffer) {
    const bytes = buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer);
    if (bytes.length < 12) {
      throw new Error("Corrupt .reflex file: header too short");
    }

    // 1. Verify 4-byte Magic Header: 'RFX1'
    const magic = String.fromCharCode(bytes[0], bytes[1], bytes[2], bytes[3]);
    if (magic !== "RFX1") {
      throw new Error(`Invalid magic header: expected 'RFX1', got '${magic}'`);
    }

    // 2. Read CRC32 and Length (Big-Endian)
    const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    const expectedCrc = view.getUint32(4, false);
    const length = view.getUint32(8, false);

    // 3. Extract and Verify Payload
    const payloadBytes = bytes.subarray(12, 12 + length);
    if (payloadBytes.length !== length) {
      throw new Error("Incomplete .reflex file: truncated payload");
    }

    const actualCrc = crc32(payloadBytes);
    if (actualCrc !== expectedCrc) {
      throw new Error(`CRC32 checksum mismatch: expected ${expectedCrc}, got ${actualCrc} (corrupt model)`);
    }

    // 4. Parse UTF-8 JSON Payload
    const decoder = new TextDecoder("utf-8");
    const jsonStr = decoder.decode(payloadBytes);
    const data = JSON.parse(jsonStr);

    return new CompiledInstinct({
      name: data.name,
      decisionType: data.decision_type,
      options: data.options || [],
      weights: data.weights || {},
      biases: data.biases || {},
      temperature: data.temperature || 1.0,
      metrics: data.metrics,
      spec: data.spec,
      compiledAt: data.compiled_at,
      version: data.version || "1.0",
    });
  }

  /**
   * Loads a .reflex file from the filesystem (Node.js / Bun).
   * @param {string} filePath
   * @returns {Promise<CompiledInstinct>}
   */
  static async fromFile(filePath) {
    if (typeof process !== "undefined" && process.versions && process.versions.node) {
      const fs = await import("node:fs/promises");
      const buffer = await fs.readFile(filePath);
      return CompiledInstinct.fromBinary(buffer);
    }
    throw new Error("CompiledInstinct.fromFile() is only supported in Node.js/Bun environments.");
  }

  /**
   * Synchronous file loader for Node.js / Bun.
   * @param {string} filePath
   * @returns {CompiledInstinct}
   */
  static fromFileSync(filePath) {
    if (typeof process !== "undefined" && process.versions && process.versions.node) {
      // Use dynamic require or import in CommonJS / ESM
      const fs = require ? require("fs") : null;
      if (fs && fs.readFileSync) {
        return CompiledInstinct.fromBinary(fs.readFileSync(filePath));
      }
    }
    throw new Error("CompiledInstinct.fromFileSync() requires synchronous Node.js fs.");
  }

  /**
   * Executes sub-20 microsecond inference on input state with calibrated probabilities.
   * @param {string} state
   * @returns {DecisionResult}
   */
  predict(state) {
    const t0 = typeof performance !== "undefined" ? performance.now() : Date.now();
    const vec = this.encoder.encode(state);

    const decisions = {};
    if (this.decisionType === "choice") {
      const logits = [];
      for (const opt of this.options) {
        const w = this.weights[opt] || new Array(vec.length).fill(0.0);
        const b = this.biases[opt] || 0.0;
        let z = b;
        for (let i = 0; i < vec.length; i++) {
          z += w[i] * vec[i];
        }
        logits.push(z);
      }

      const probs = softmax(logits, this.temperature);
      const dist = {};
      let bestIdx = 0;
      let maxP = -1;
      for (let i = 0; i < this.options.length; i++) {
        const p = Math.round(probs[i] * 10000) / 10000;
        dist[this.options[i]] = p;
        if (probs[i] > maxP) {
          maxP = probs[i];
          bestIdx = i;
        }
      }

      const bestOpt = this.options[bestIdx] || "";
      decisions.choice = new Choice({
        instructions: this.spec?.prompt || this.name,
        options: this.options,
        selected: bestOpt,
        distribution: dist,
      });
    } else if (this.decisionType === "noul") {
      const w = this.weights["noul"] || new Array(vec.length).fill(0.0);
      const b = this.biases["noul"] || 0.0;
      let z = b;
      for (let i = 0; i < vec.length; i++) {
        z += w[i] * vec[i];
      }
      const prob = sigmoid(z / this.temperature);
      decisions.noul = new Noul({
        instructions: this.spec?.prompt || this.name,
        probability: Math.round(prob * 10000) / 10000,
      });
    } else {
      const w = this.weights["score"] || new Array(vec.length).fill(0.0);
      const b = this.biases["score"] || 0.0;
      let z = b;
      for (let i = 0; i < vec.length; i++) {
        z += w[i] * vec[i];
      }
      const normVal = sigmoid(z / this.temperature);
      const scoreVal = Math.round((1.0 + normVal * 9.0) * 100) / 100;
      decisions.score = new Score({
        instructions: this.spec?.prompt || this.name,
        score: scoreVal,
        confidence: 0.95,
      });
    }

    const t1 = typeof performance !== "undefined" ? performance.now() : Date.now();
    const latencyMs = Math.round((t1 - t0) * 1000) / 1000;

    return new DecisionResult({
      decisions,
      latencyMs,
      backend: `compiled:${this.name}`,
      inputTokens: state.split(/\s+/).filter(Boolean).length,
      outputTokens: 0,
      costUsd: 0.0,
    });
  }
}

File: reflex/compiler.py
"""
Reflex Prompt-to-Instinct Compiler & Calibration Pipeline (Phase 24).
Distills 2,000-token LLM system prompts into portable, sub-50us local decision heads (.reflex artifacts).
Zero external dependencies (Python standard library only).
"""

from __future__ import annotations
from dataclasses import dataclass, field
import hashlib
import json
import math
import os
import re
import struct
import time
import zlib
from typing import Any, Callable, Dict, List, Optional, Tuple, Union

from reflex.embeddings import SemanticVectorEncoder, cosine_similarity
from reflex.primitives import Choice, Noul, Score, DecisionResult


MAGIC_HEADER = b"RFX1"  # Reflex Compiled Model Format v1


def _sigmoid(z: float) -> float:
    clamped = max(-30.0, min(30.0, z))
    return 1.0 / (1.0 + math.exp(-clamped))


def _softmax(logits: List[float], temperature: float = 1.0) -> List[float]:
    if not logits:
        return []
    temp = max(0.01, temperature)
    scaled = [x / temp for x in logits]
    max_l = max(scaled)
    exps = [math.exp(max(-30.0, min(30.0, x - max_l))) for x in scaled]
    sum_exps = sum(exps)
    if sum_exps <= 0.0:
        return [1.0 / len(logits)] * len(logits)
    return [e / sum_exps for e in exps]


@dataclass
class PromptSpec:
    """
    Specification for compiling a prompt into an instinct model.
    """
    prompt: str
    decision_type: str = "choice"  # "choice" | "noul" | "score"
    options: List[str] = field(default_factory=list)
    guidelines: Dict[str, str] = field(default_factory=dict)
    few_shot_examples: List[Dict[str, Any]] = field(default_factory=list)
    name: str = "compiled_model"

    def to_dict(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "prompt": self.prompt,
            "decision_type": self.decision_type,
            "options": self.options,
            "guidelines": self.guidelines,
            "few_shot_examples": self.few_shot_examples,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "PromptSpec":
        return cls(
            name=data.get("name", "compiled_model"),
            prompt=data.get("prompt", ""),
            decision_type=data.get("decision_type", "choice"),
            options=data.get("options", []),
            guidelines=data.get("guidelines", {}),
            few_shot_examples=data.get("few_shot_examples", []),
        )


@dataclass
class CalibrationMetrics:
    """Performance and calibration validation metrics of a compiled model."""
    accuracy: float
    brier_score: float
    ece: float
    total_samples: int
    training_time_ms: float

    def to_dict(self) -> Dict[str, Any]:
        return {
            "accuracy": round(self.accuracy, 4),
            "brier_score": round(self.brier_score, 4),
            "ece": round(self.ece, 4),
            "total_samples": self.total_samples,
            "training_time_ms": round(self.training_time_ms, 2),
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "CalibrationMetrics":
        return cls(
            accuracy=data.get("accuracy", 0.0),
            brier_score=data.get("brier_score", 0.0),
            ece=data.get("ece", 0.0),
            total_samples=data.get("total_samples", 0),
            training_time_ms=data.get("training_time_ms", 0.0),
        )


class CompiledInstinct:
    """
    Portable, self-contained compiled decision model evaluating in <50 microseconds ($0.05ms).
    """

    def __init__(
        self,
        name: str,
        decision_type: str,
        options: List[str],
        weights: Dict[str, List[float]],
        biases: Dict[str, float],
        temperature: float = 1.0,
        metrics: Optional[CalibrationMetrics] = None,
        spec: Optional[PromptSpec] = None,
        compiled_at: Optional[str] = None,
        version: str = "1.0",
    ):
        self.name = name
        self.decision_type = decision_type
        self.options = options
        self.weights = weights
        self.biases = biases
        self.temperature = max(0.01, temperature)
        self.metrics = metrics or CalibrationMetrics(1.0, 0.0, 0.0, 0, 0.0)
        self.spec = spec
        self.compiled_at = compiled_at or time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        self.version = version
        self._encoder = SemanticVectorEncoder()
        from reflex.simd import get_simd_engine
        self._simd = get_simd_engine()
        self.quantization = "fp32"
        self.quantized_weights = {}
        self.scales = {}

    def quantize(self, mode: str = "int8") -> "CompiledInstinct":
        """
        Quantizes FP32 weights to INT8 to accelerate inference and reduce memory footprint.
        """
        if mode != "int8":
            raise ValueError(f"Unsupported quantization mode: {mode}")
        self.quantization = "int8"
        self.quantized_weights = {}
        self.scales = {}
        for key, w in self.weights.items():
            q_bytes, scale = self._simd.quantize_i8(w)
            self.quantized_weights[key] = q_bytes
            self.scales[key] = scale
        return self

    def predict(self, state: str) -> DecisionResult:
        """
        Executes sub-50 microsecond inference on input state with calibrated probabilities.
        """
        t0 = time.perf_counter()
        vec = self._encoder.encode(state)

        decisions = {}
        if self.decision_type == "choice":
            logits = []
            if self.quantization == "int8" and self.quantized_weights:
                q_vec, s_vec = self._simd.quantize_i8(vec)
                for opt in self.options:
                    w_q = self.quantized_weights.get(opt, b"")
                    w_s = self.scales.get(opt, 1.0)
                    b = self.biases.get(opt, 0.0)
                    z = self._simd.dot_product_i8(w_q, w_s, q_vec, s_vec) + b
                    logits.append(z)
            else:
                for opt in self.options:
                    w = self.weights.get(opt, [0.0] * len(vec))
                    b = self.biases.get(opt, 0.0)
                    z = self._simd.dot_product_f32(w, vec) + b
                    logits.append(z)

            probs = _softmax(logits, temperature=self.temperature)
            dist = {opt: round(p, 4) for opt, p in zip(self.options, probs)}
            best_idx = max(range(len(probs)), key=lambda i: probs[i]) if probs else 0
            best_opt = self.options[best_idx] if self.options else ""

            choice_obj = Choice(
                instructions=self.spec.prompt if self.spec else self.name,
                options=self.options,
                selected=best_opt,
                distribution=dist,
            )
            decisions["choice"] = choice_obj

        elif self.decision_type == "noul":
            w = self.weights.get("noul", [0.0] * len(vec))
            b = self.biases.get("noul", 0.0)
            z = (self._simd.dot_product_f32(w, vec) + b) / self.temperature
            prob = _sigmoid(z)

            noul_obj = Noul(
                instructions=self.spec.prompt if self.spec else self.name,
                probability=round(prob, 4),
            )
            decisions["noul"] = noul_obj

        else:
            w = self.weights.get("score", [0.0] * len(vec))
            b = self.biases.get("score", 0.0)
            z = (self._simd.dot_product_f32(w, vec) + b) / self.temperature
            norm_val = _sigmoid(z)
            score_val = round(1.0 + (norm_val * 9.0), 2)  # Scale 1.0 - 10.0

            score_obj = Score(
                instructions=self.spec.prompt if self.spec else self.name,
                score=score_val,
                confidence=0.95,
            )
            decisions["score"] = score_obj

        latency_ms = (time.perf_counter() - t0) * 1000.0

        return DecisionResult(
            decisions=decisions,
            latency_ms=round(latency_ms, 3),
            backend=f"compiled:{self.name}",
            input_tokens=len(state.split()),
            output_tokens=0,
            cost_usd=0.0,
        )

    def save(self, path: str) -> None:
        """
        Serializes the compiled model into a portable .reflex binary artifact.
        Format: [MAGIC 4B][CRC32 4B][JSON LENGTH 4B][JSON PAYLOAD UTF-8]
        """
        payload_data = {
            "name": self.name,
            "decision_type": self.decision_type,
            "options": self.options,
            "weights": self.weights,
            "biases": self.biases,
            "temperature": self.temperature,
            "metrics": self.metrics.to_dict(),
            "spec": self.spec.to_dict() if self.spec else None,
            "compiled_at": self.compiled_at,
            "version": self.version,
        }
        json_bytes = json.dumps(payload_data, separators=(",", ":")).encode("utf-8")
        crc = zlib.crc32(json_bytes) & 0xFFFFFFFF
        length = len(json_bytes)

        header = struct.pack(">4sII", MAGIC_HEADER, crc, length)
        with open(path, "wb") as f:
            f.write(header)
            f.write(json_bytes)

    @classmethod
    def load(cls, path: str) -> "CompiledInstinct":
        """
        Loads and verifies a .reflex model artifact from disk.
        """
        with open(path, "rb") as f:
            header = f.read(12)
            if len(header) < 12:
                raise ValueError("Corrupt .reflex file: header too short")
            magic, expected_crc, length = struct.unpack(">4sII", header)
            if magic != MAGIC_HEADER:
                raise ValueError(f"Invalid magic header: expected {MAGIC_HEADER}, got {magic}")

            json_bytes = f.read(length)
            if len(json_bytes) != length:
                raise ValueError("Incomplete .reflex file: truncated payload")

            actual_crc = zlib.crc32(json_bytes) & 0xFFFFFFFF
            if actual_crc != expected_crc:
                raise ValueError(f"CRC32 checksum mismatch: expected {expected_crc}, got {actual_crc} (corrupt model)")

        data = json.loads(json_bytes.decode("utf-8"))
        spec = PromptSpec.from_dict(data["spec"]) if data.get("spec") else None
        metrics = CalibrationMetrics.from_dict(data.get("metrics", {}))

        return cls(
            name=data["name"],
            decision_type=data["decision_type"],
            options=data["options"],
            weights=data["weights"],
            biases=data["biases"],
            temperature=data.get("temperature", 1.0),
            metrics=metrics,
            spec=spec,
            compiled_at=data.get("compiled_at"),
            version=data.get("version", "1.0"),
        )


class SyntheticDataGenerator:
    """
    Pure-Python synthetic training state generator from prompt guidelines and options.
    Generates balanced, diverse domain queries without needing external API calls.
    """

    def __init__(self):
        self.templates = [
            "Customer states: {content}",
            "User query: {content}",
            "Ticket description: {content}",
            "Regarding {content}",
            "Urgent request: {content}",
            "Question about {content}",
            "{content}",
        ]

    def generate_samples_for_option(
        self,
        option: str,
        guideline: str,
        count: int = 30,
    ) -> List[str]:
        """Generates diverse training utterances for a given candidate option."""
        # Extract keywords and concepts from guideline and option name
        clean_words = [w for w in re.findall(r"\w+", f"{option} {guideline}".lower()) if len(w) > 3]
        if not clean_words:
            clean_words = [option]

        samples = []
        for i in range(count):
            tmpl = self.templates[i % len(self.templates)]
            # Construct varied sentences using guideline phrases
            kw = clean_words[i % len(clean_words)]
            alt_kw = clean_words[(i + 1) % len(clean_words)]
            
            variants = [
                f"Please help me with {kw} and {alt_kw}.",
                f"I have an issue concerning my {kw}.",
                f"Need immediate assistance with {option}: {guideline}.",
                f"How do I update or manage my {kw}?",
                f"Problem report: {kw} is failing.",
                f"Inquiry regarding {alt_kw} policy.",
            ]
            content = variants[i % len(variants)]
            samples.append(tmpl.format(content=content))

        return samples


class InstinctCompiler:
    """
    High-level compiler distilling prompt specifications into calibrated .reflex models.
    """

    def __init__(self):
        self.encoder = SemanticVectorEncoder()
        self.generator = SyntheticDataGenerator()

    def compile(
        self,
        spec: PromptSpec,
        samples_per_class: int = 35,
        epochs: int = 40,
        lr: float = 0.08,
    ) -> CompiledInstinct:
        """
        Compiles a PromptSpec into a calibrated CompiledInstinct model.
        """
        t0 = time.perf_counter()
        options = spec.options or ["negative", "positive"]

        # 1. Generate or aggregate training dataset
        dataset: List[Tuple[str, str]] = []
        for opt in options:
            guideline = spec.guidelines.get(opt, spec.prompt)
            samples = self.generator.generate_samples_for_option(opt, guideline, count=samples_per_class)
            for s in samples:
                dataset.append((s, opt))

        # Include few-shot examples if provided
        for ex in spec.few_shot_examples:
            txt = ex.get("text") or ex.get("state") or ""
            label = ex.get("label") or ex.get("selected") or options[0]
            if txt and label in options:
                dataset.append((txt, label))

        # 2. Vector encoding
        X: List[List[float]] = [self.encoder.encode(text) for text, _ in dataset]
        dim = len(X[0]) if X else 384

        # 3. Model parameters initialization
        weights: Dict[str, List[float]] = {opt: [0.0] * dim for opt in options}
        biases: Dict[str, float] = {opt: 0.0 for opt in options}

        # 4. Multi-class Logistic Regression with momentum & L2 penalty
        v_w: Dict[str, List[float]] = {opt: [0.0] * dim for opt in options}
        v_b: Dict[str, float] = {opt: 0.0 for opt in options}
        momentum = 0.9
        l2_reg = 0.001

        num_samples = len(dataset)
        for epoch in range(epochs):
            cur_lr = lr * (1.0 - (epoch / (epochs * 1.5)))

            for (text, label), x in zip(dataset, X):
                # Forward pass: compute logits and softmax probabilities
                logits = [
                    sum(wi * xi for wi, xi in zip(weights[opt], x)) + biases[opt]
                    for opt in options
                ]
                probs = _softmax(logits)

                # Compute gradients: dL/dz_i = p_i - y_i
                for idx, opt in enumerate(options):
                    y_i = 1.0 if opt == label else 0.0
                    grad_z = probs[idx] - y_i

                    # Update bias
                    v_b[opt] = momentum * v_b[opt] - cur_lr * grad_z
                    biases[opt] += v_b[opt]

                    # Update weights
                    w_opt = weights[opt]
                    v_w_opt = v_w[opt]
                    for d in range(dim):
                        grad_w = grad_z * x[d] + l2_reg * w_opt[d]
                        v_w_opt[d] = momentum * v_w_opt[d] - cur_lr * grad_w
                        w_opt[d] += v_w_opt[d]

        # 5. Validation metrics computation
        correct = 0
        brier_sum = 0.0
        confidences = []
        accuracies = []

        for (text, label), x in zip(dataset, X):
            logits = [
                sum(wi * xi for wi, xi in zip(weights[opt], x)) + biases[opt]
                for opt in options
            ]
            probs = _softmax(logits)
            pred_idx = max(range(len(probs)), key=lambda i: probs[i])
            pred_opt = options[pred_idx]

            is_correct = 1.0 if pred_opt == label else 0.0
            if pred_opt == label:
                correct += 1

            for idx, opt in enumerate(options):
                y = 1.0 if opt == label else 0.0
                brier_sum += (probs[idx] - y) ** 2

            confidences.append(probs[pred_idx])
            accuracies.append(is_correct)

        accuracy = correct / max(1, num_samples)
        brier_score = brier_sum / (max(1, num_samples) * len(options))

        # Expected Calibration Error (ECE) with 5 bins
        ece = 0.0
        num_bins = 5
        for b in range(num_bins):
            low, high = b / num_bins, (b + 1) / num_bins
            bin_indices = [i for i, c in enumerate(confidences) if low <= c < high]
            if bin_indices:
                bin_acc = sum(accuracies[i] for i in bin_indices) / len(bin_indices)
                bin_conf = sum(confidences[i] for i in bin_indices) / len(bin_indices)
                ece += (len(bin_indices) / num_samples) * abs(bin_acc - bin_conf)

        t_elapsed_ms = (time.perf_counter() - t0) * 1000.0

        metrics = CalibrationMetrics(
            accuracy=accuracy,
            brier_score=brier_score,
            ece=ece,
            total_samples=num_samples,
            training_time_ms=t_elapsed_ms,
        )

        return CompiledInstinct(
            name=spec.name,
            decision_type=spec.decision_type,
            options=options,
            weights=weights,
            biases=biases,
            temperature=1.0,
            metrics=metrics,
            spec=spec,
        )
